include/zephyr/device.h

```diff
@@ -737,12 +768,157 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
 	return z_device_is_ready(dev);
 }
 
+/**
+ * @brief Define device subsystem
+ *
+ * @details Places device subsystem implementation in appropriate
+ * section in ROM.
+ *
+ * @example
+ *
+ * DEVICE_SUBSYS(rtc, rtc_emul_driver_api) = {
+ *         get_time = rtc_emul_get_time;
+ * };
+ *
+ * @param _subsys Device subsystem [rtc, i2c, ...]
+ * @param _varname Variable name
+ */
+#define DEVICE_SUBSYS(_subsys, _varname) \
+	const STRUCT_SECTION_ITERABLE(_subsys##_driver_api, _varname)
+
+#if CONFIG_DEVICE_LIST_SUBSYSTEMS
+
+/**
+ * @brief Get pointer to struct device subsystem
+ *
+ * @param _dev Pointer to struct device
+ * @param _subsys Device subsystem [rtc, i2c, ...]
+ *
+ * @example
+ *
+ * const struct rtc_driver_api *api = DEVICE_SUBSYS_GET(dev, rtc);
+ *
+ * @note Leaves scope returning -ENOSYS if struct device does not
+ * implement specified subsystem.
+ *
+ * @note Subsystem must be defined using DEVICE_SUBSYS() macro
+ *
+ * @return Pointer to implemented struct device subsystem
+ */
+#define DEVICE_SUBSYS_GET(_dev, _subsys)					\
+	(const struct _subsys##_driver_api *)_dev->api;				\
+	STRUCT_SECTION_START_EXTERN(_subsys##_driver_api);			\
+	STRUCT_SECTION_END_EXTERN(_subsys##_driver_api);			\
+	do {									\
+		if (api > TYPE_SECTION_END(_subsys##_driver_api)) {		\
+			api = (const struct _subsys##_driver_api *)		\
+			      (((uintptr_t *)api) + 1);				\
+										\
+			continue;						\
+		}								\
+										\
+		if (api >= TYPE_SECTION_START(_subsys##_driver_api)) {		\
+			break;							\
+		}								\
+										\
+		if (api == NULL) {						\
+			return -ENOSYS;						\
```

Comment:
a macro returning?

Reply:
@gmarull State sharing (from subsystem listing) is intended for devices like I3C, which inherit existing APIs. I had to use existing APIs for the testing :)  It could also be used for devices like UARTs, which have three quite different APIs in one, these could be single APIs instead (poll, isr, async), which is a bit cleaner IMO and potentially  more memory efficient.

I agree we should have true device classes, giving us more generic parent/subdevice management to handle PM and access synchronization, but that's a topic of its own I think :)